### PR TITLE
Disable landing pads within compiler crates

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
Generally speaking if the compiler panics we just need to emit the ICE
message, there's no particular need to run drop.

Instead of compiling the world with -Cpanic=abort however we use
-Zno-landing-pads to remain compatible with -Cpanic=unwind crates (e.g.,
proc macros) which want to be linked with unwinding because they use
panics for error handling. It's unclear whether this is simply a
limitation of the panic implementation strategy.

This nets wins of 3-6% on most benchmarks.

### CODE REVIEW

**Comment by nnethercote** [comment]
@Mark-Simulacrum: Nice improvements! What prompted you to try this?

Also, should the code have some comments explaining why these settings were chosen?

**Comment by Mark-Simulacrum** [comment]
Working on [another project](https://github.com/Mark-Simulacrum/ecap-rs) of mine I was thinking about panic handling so decided to try this -- thought it might net some wins. Initially tried panic=abort but that (which is possibly better than this solution) doesn't quite work today I think, though I didn't investigate in detail.

**Review by bjorn3 — CHANGES_REQUESTED** [review]

File: src/bootstrap/compile.rs
Comment (Mark-Simulacrum) [reply src/bootstrap/compile.rs]: Ah, I think I forgot it after changing the env var in the other file. I'll run this through official perf just to make sure though.

**Comment by alexcrichton** [comment]
@rust-timer build 4c9951f5c153fe4ce033b18371beeb6a48e6a728

**Comment by Mark-Simulacrum** [comment]
Hm, official perf is less impressive than my local results (only 3% wins).

I've not investigated the UI failures yet, I'd like to get signoff from relevant folks (perhaps we should FCP compiler team?) before putting much more time into this.

**Comment by nnethercote** [comment]
> Hm, official perf is less impressive than my local results (only 3% wins).

The breadth of the improvements is impressive, though.

**Comment by alexcrichton** [comment]
The test failures here are an example of why I think it may be technically infeasible to land this strategy at this time, with the corrected patch now is it still sure that failing rustdoc tests, procedural macros, and compile errors in general all still work?

**Comment by nikomatsakis** [comment]
I personally think this sounds like a fine thing to do, though I'd prefer if we could do it with `-Cpanic=abort`, but I guess that'd be a lot more work. That said, I'm not sure if it's worth investing a lot of time into it at this point (vs filing an issue and revisiting later), given the relatively minor gains and the sheer number of things in flight due to the Edition.

**Comment by Mark-Simulacrum** [comment]
I agree that holding off until after the edition and looking at properly doing this via -Cpanic=abort is the best approach here; closing this for the time being.

**Pull request closed without merging**
